### ISSUE TITLE
WalkingLady.gltf may be incorrect

### ISSUE THREAD

**Issue by cx20** [KhronosGroup/glTF-Sample-Models#83]
I attempted to display the [WalkingLady.gltf](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/WalkingLady/glTF/WalkingLady.gltf) model of glTF 2.0 with the latest Three.js + glTF Loader.

[Three.js r86 + glTF Loader result:](https://cdn.rawgit.com/cx20/gltf-test/c41c05d1f39e4bc97d4df37d9723e8e5863b4424/examples/threejs/index.html?model=WalkingLady&scale=1)
![image](https://user-images.githubusercontent.com/1833979/27387681-5724d748-56d4-11e7-9144-974de5d1104b.png)

I could not judge whether it is a model (or converter) problem or a viewer problem.

Related: https://github.com/KhronosGroup/glTF/issues/685

**Comment by javagl** [comment]
Just a (maybe obvious) guess: The most important change between 1.0 and 2.0 in this regard was that the `bindShapeMatrix` is supposed to me "baked" into the geometry - I'm not sure if this is already considered and/or tested thoroughly, but if it was mising, then this **could** explain this bug. (I'd like to examine this further, but the bandwidth here is limited...)

**Comment by cx20** [comment]
I also tried glTF 2.0 Loader other than Three.js.

[Grimoire.js + glTF Loader result:](https://cdn.rawgit.com/cx20/gltf-test/0ece75c19338a1e699e7c2e80d115bd76eb2df4d/examples/grimoiregl/index.html?model=WalkingLady&scale=1)
![image](https://user-images.githubusercontent.com/1833979/27504901-7b4886d2-58ce-11e7-9a9b-d5b40eee77df.png)
Grimoire.js seems to have the same display result as Three.js.

[Babylon.js + glTF Loader result:](https://cdn.rawgit.com/cx20/gltf-test/0ece75c19338a1e699e7c2e80d115bd76eb2df4d/examples/babylonjs/index.html?model=WalkingLady&scale=1)
![image](https://user-images.githubusercontent.com/1833979/27504944-30ad81da-58cf-11e7-8e81-ce0588f43587.png)
Since Babylon.js was not able to display WalkingLady.gltf correctly even with glTF 1.0 Loader, there is a possibility that another problem may be hidden.

**Comment by cx20** [comment]
I tried the simple skinning animation model below.
[EmaSimpleSkin.dae](https://github.com/cx20/jsdo-static-contents/blob/master/models/collada/EmaSimpleSkin/EmaSimpleSkin.dae) : This is a COLLADA file exported from Maya using OpenCOLLADA 2016. This sample was donated by @emadurandal
I converted it into glTF 2.0 format using [COLLADA2GLTF](https://github.com/KhronosGroup/COLLADA2GLTF).

The following is the display result by Three.js. As animation movement the glTF 1.0 model is more correct.
Because WalkingLady.dae has also been exported from Maya, I think that it includes the same problem.

[Three.js + glTF 1.0 model result:](http://jsdo.it/cx20/coRj)
![emasimpleskin_threejs_gltf1](https://user-images.githubusercontent.com/1833979/27516485-f37ef488-59f5-11e7-8467-48a024c393c7.gif)

[Three.js + glTF 2.0 model result:](http://jsdo.it/cx20/q0cx)
![emasimpleskin_threejs_gltf2](https://user-images.githubusercontent.com/1833979/27516487-fb2fb85c-59f5-11e7-85da-1fde932e3605.gif)

**Comment by cx20** [comment]
I also tried Blender 2.78c and glTF 2.0 Explorter.
First I tried to display [EmaSimpleSkin.dae](https://github.com/cx20/jsdo-static-contents/blob/master/models/collada/EmaSimpleSkin/EmaSimpleSkin.dae) on Blender.
![emasimpleskin_blender_gltf2_20170629_002](https://user-images.githubusercontent.com/1833979/27683813-5785f166-5d02-11e7-966d-022e25e9bf5c.gif)
Next, I tried to display the glTF 2.0 file exported from Blender in Three.js.
[Three.js + glTF 2.0 model result:](http://jsdo.it/cx20/yfvq)
![emasimpleskin_threejs_gltf2_20170629](https://user-images.githubusercontent.com/1833979/27683862-82afeb9e-5d02-11e7-8406-f27db34ff958.gif)

I think that it is not a viewer but a converter (or exporter) issue so I think that it is necessary for each tool side to deal with Issue.
- Blender glTF exporter
- COLLADA2GLTF

**Comment by donmccurdy** [comment]
Some users are getting confused by this model; I've had people ask questions and file bugs about it (see two linked issues above).

Perhaps we should delete the glTF file (but keep the COLLADA source model) until we have something valid? Or put a large warning in the README.md? Unless someone has bandwidth to look into the issue itself in the near future; unfortunately I do not.

**Comment by emadurandal** [comment]
@donmccurdy 

I have created this WalkingLady model using Maya.

At the time, my GLBoost library and javagl/Jgltf succeeded in playback first, then later Three.js etc. also succeeded in playing this model with thier 1.0 compatible loader.

I think that this model is logically correct as animation as glTF 1.0, but I am not confident that I can explain details to prove it.

I do not want to confuse the community, so I agree to dismiss this model.

　

(Following is my personal opinion, Sorry for off track)

This problem seems to be a compatibility problem of the combination of COLLADA2GLTF tool and Maya.

It is not desirable that the playing result differs depending on the production workflow of the glTF file. However, as the implementation is different, I think that it can happen as a real problem.

Does this community show something about officially recommended production workflow for creating glTF files?

It may be good to have some official view on the recommended workflow at the moment.

(In particular, it seems that there is not much experience in the community for models made in Maya.)

**Comment by javagl** [comment]
**tl;dr** : The model indeed seems to be wrong, and should be removed until we're sure that it is valid.

---

When a model appears to be "wrong", it's sometimes hard to spot the culprit: It could be...

- a "bug" in the original model (e.g. the original Maya model)
- a "bug" in the COLLADA model (that was exported from Maya)
- a "bug" in COLLADA2GLTF
- a bug in the viewer

The term "bug" is in "quotes" above, meaning that the model/file may not really be wrong or invalid, but might just use a certain representation of things that is not yet covered or anticipated by the downstream pipeline. E.g. COLLADA2GLTF or a certain viewer might have some expectation about the input structure, and does not know that there are other representations that are still valid.

---

The WalkingLady model is nearly "famous" for throwing out some renderers - although sometimes for [very subtle reasons](https://github.com/AnalyticalGraphicsInc/cesium/issues/4722#issuecomment-269502004) ;-) And I think it's good to have a model that obviously uses a **valid** representation of something that still differs from what other models contain. 

For example, one of the main differences between the WalkingLady and basically **all** other models is: The `bindShapeMatrix` is **not** the identity matrix.

Viewers that are tested with the other models might render them all properly, and glitch out with the WalkingLady, assuming that it contains a bug. But it does not (necessarily) contain a bug. I just uses some feature that other models don't use.

---

However, I did a short investigation of this. 

<sub>This is based on the [2.0 branch of JglTF](https://github.com/javagl/JglTF/tree/2.0), which unfortunately is not yet "finished", and I had to insert some crude hacks (locally) to render the model at all (this PBR thing... it's difficult). So the following should be taken with a grain of salt. But...</sub>

... from what I have observed, the model is indeed wrong.

This is the initial result of trying to render it:

![walkinglady2_initial](https://user-images.githubusercontent.com/5597569/30441500-edb2c9a8-9979-11e7-9874-6caf8a6ff7a7.png)

The following is the result that was created from manually baking the `bindShapeMatrix` into the `POSITIONS`. As mentioned above, one of the main differences between glTF 1.0 and 2.0 in this regard is that the `bindShapeMatrix` does no longer exist, and it is assumed that it is already pre-multiplied with the `POSITIONS` : 

![walkinglady2_bakedbindshape](https://user-images.githubusercontent.com/5597569/30441606-362da504-997a-11e7-9f71-fa0d3c4e1e8e.png)

Then I had a look at the `inverseBindMatrices` accessor, using the [jgltf-browser](https://github.com/javagl/JglTF/tree/master/jgltf-browser). I compared the `inverseBindMatrices` of the WalkingLady 1.0 and the 2.0 version, and noticed that they are different.  So I manually copied the `inverseBindMatrices` from the 1.0 version into the 2.0 version (hacky, hacky...). The result is shown here:

![walkinglady2_oldibm](https://user-images.githubusercontent.com/5597569/30441797-ba7d42a6-997a-11e7-82bc-cffe401defcd.png)

Finally, I did **both** : Using the `inverseBindMatrices` from the 1.0 model, **and** pre-multiplied the `POSITIONS` with the `bindShapeMatrix`. And this is the result:

![walkinglady2_bakedbindshape_and_oldibm](https://user-images.githubusercontent.com/5597569/30441839-d9436b7a-997a-11e7-9c6d-d055d17a367a.png)

---

So as far as I can tell, the problem is that in the exported model

- the `bindShapeMatrix` is **not** premultiplied with the geometry and
- the `inverseBindMatrices` are plainly wrong

Having a closer look at the `inverseBindMatrices` reveals that the problems may be connected: 

These are the `inverseBindMatrices` of version 1.0:

    [(   0.00000,   -1.00000,    0.00000,    0.00000,    1.00000,    0.00000,    0.00000,    0.00000,    0.00000,    0.00000,    1.00000,    0.00000,   -1.05487,    0.00000,    0.00000,    1.00000), 
     (   0.00000,   -1.00000,    0.00000,    0.00000,    1.00000,    0.00000,    0.00000,    0.00000,    0.00000,    0.00000,    1.00000,    0.00000,   -1.38978,    0.00000,    0.00000,    1.00000), 
     (   0.00442,   -0.99999,    0.00000,    0.00000,    0.99999,    0.00442,    0.00000,    0.00000,    0.00000,    0.00000,    1.00000,    0.00000,   -1.45982,   -0.00646,    0.00000,    1.00000), 
     (   1.00000,    0.00000,    0.00000,    0.00000,    0.00000,    1.00000,    0.00000,    0.00000,    0.00000,    0.00000,    1.00000,    0.00000,   -0.00109,   -1.70719,    0.00000,    1.00000), 
     (   0.78087,    0.62470,    0.00000,    0.00000,   -0.62470,    0.78087,    0.00000,    0.00000,    0.00000,    0.00000,    1.00000,    0.00000,    0.77215,   -1.15647,    0.01045,    1.00000), 
     (   0.75323,    0.65775,    0.00000,    0.00000,   -0.65775,    0.75323,    0.00000,    0.00000,    0.00000,    0.00000,    1.00000,    0.00000,    0.55518,   -1.13360,    0.01045,    1.00000), 
     (   0.00000,   -1.00000,    0.00000,    0.00000,    1.00000,    0.00000,    0.00000,    0.00000,    0.00000,    0.00000,    1.00000,    0.00000,   -1.01547,    0.56058,    0.01045,    1.00000), 
     (  -0.78086,   -0.62471,    0.00000,    0.00000,   -0.62471,    0.78086,    0.00000,    0.00000,    0.00000,    0.00000,   -1.00000,    0.00000,    0.77217,   -1.15646,   -0.01045,    1.00000), 
     (  -0.75324,   -0.65774,    0.00000,    0.00000,   -0.65774,    0.75324,    0.00000,    0.00000,    0.00000,    0.00000,   -1.00000,    0.00000,    0.55517,   -1.13361,   -0.01045,    1.00000), 
     (   0.00000,   -1.00000,    0.00000,    0.00000,   -1.00000,    0.00000,    0.00000,    0.00000,    0.00000,    0.00000,   -1.00000,    0.00000,    1.01547,   -0.56058,   -0.01045,    1.00000), 
     (  -0.05907,   -0.76017,   -0.64704,    0.00000,   -0.99699,    0.07747,    0.00000,    0.00000,    0.05012,    0.64509,   -0.76246,    0.00000,    0.95732,   -0.01062,    0.05395,    1.00000), 
     (  -0.01038,   -0.06613,    0.99776,    0.00000,   -0.98791,    0.15501,    0.00000,    0.00000,   -0.15467,   -0.98570,   -0.06694,    0.00000,    0.52785,   -0.05731,   -0.05664,    1.00000), 
     (   0.00000,   -1.00000,    0.00000,    0.00000,    1.00000,    0.00000,    0.00000,    0.00000,    0.00000,    0.00000,    1.00000,    0.00000,   -0.00936,    0.05273,    0.06020,    1.00000), 
     (   0.05907,    0.76017,   -0.64704,    0.00000,   -0.99699,    0.07747,    0.00000,    0.00000,    0.05012,    0.64509,    0.76246,    0.00000,    0.95732,   -0.01062,   -0.05395,    1.00000), 
     (   0.01038,    0.06613,    0.99776,    0.00000,   -0.98791,    0.15501,    0.00000,    0.00000,   -0.15467,   -0.98570,    0.06694,    0.00000,    0.52786,   -0.05731,    0.05664,    1.00000), 
     (   0.00000,   -1.00000,    0.00000,    0.00000,   -1.00000,    0.00000,    0.00000,    0.00000,    0.00000,    0.00000,   -1.00000,    0.00000,    0.00936,   -0.05273,   -0.06020,    1.00000)]

These are the `inverseBindMatrices` of version 2.0:

    [(   0.00000,   -0.25000,    0.00000,    0.00000,    0.25000,    0.00000,    0.00000,    0.00000,    0.00000,    0.00000,    0.25000,    0.00000,   -0.26372,    1.29397,    0.00000,    1.00000), 
     (   0.00000,   -0.25000,    0.00000,    0.00000,    0.25000,    0.00000,    0.00000,    0.00000,    0.00000,    0.00000,    0.25000,    0.00000,   -0.34745,    1.29397,    0.00000,    1.00000), 
     (   0.00111,   -0.25000,    0.00000,    0.00000,    0.25000,    0.00111,    0.00000,    0.00000,    0.00000,    0.00000,    0.25000,    0.00000,   -0.36495,    1.29235,    0.00000,    1.00000), 
     (   0.25000,    0.00000,    0.00000,    0.00000,    0.00000,    0.25000,    0.00000,    0.00000,    0.00000,    0.00000,    0.25000,    0.00000,   -0.00027,    0.86717,    0.00000,    1.00000), 
     (   0.19522,    0.15617,    0.00000,    0.00000,   -0.15617,    0.19522,    0.00000,    0.00000,    0.00000,    0.00000,    0.25000,    0.00000,    0.19304,    1.00485,    0.00261,    1.00000), 
     (   0.18831,    0.16444,    0.00000,    0.00000,   -0.16444,    0.18831,    0.00000,    0.00000,    0.00000,    0.00000,    0.25000,    0.00000,    0.13880,    1.01057,    0.00261,    1.00000), 
     (   0.00000,   -0.25000,    0.00000,    0.00000,    0.25000,    0.00000,    0.00000,    0.00000,    0.00000,    0.00000,    0.25000,    0.00000,   -0.25387,    1.43411,    0.00261,    1.00000), 
     (  -0.19521,   -0.15618,    0.00000,    0.00000,   -0.15618,    0.19521,    0.00000,    0.00000,    0.00000,    0.00000,   -0.25000,    0.00000,    0.19304,    1.00485,   -0.00261,    1.00000), 
     (  -0.18831,   -0.16444,    0.00000,    0.00000,   -0.16444,    0.18831,    0.00000,    0.00000,    0.00000,    0.00000,   -0.25000,    0.00000,    0.13879,    1.01056,   -0.00261,    1.00000), 
     (   0.00000,   -0.25000,    0.00000,    0.00000,   -0.25000,    0.00000,    0.00000,    0.00000,    0.00000,    0.00000,   -0.25000,    0.00000,    0.25387,    1.15382,   -0.00261,    1.00000), 
     (  -0.01477,   -0.19004,   -0.16176,    0.00000,   -0.24925,    0.01937,    0.00000,    0.00000,    0.01253,    0.16127,   -0.19061,    0.00000,    0.23933,    1.29131,    0.01349,    1.00000), 
     (  -0.00259,   -0.01653,    0.24944,    0.00000,   -0.24698,    0.03875,    0.00000,    0.00000,   -0.03867,   -0.24642,   -0.01674,    0.00000,    0.13196,    1.27964,   -0.01416,    1.00000), 
     (   0.00000,   -0.25000,    0.00000,    0.00000,    0.25000,    0.00000,    0.00000,    0.00000,    0.00000,    0.00000,    0.25000,    0.00000,   -0.00234,    1.30715,    0.01505,    1.00000), 
     (   0.01477,    0.19004,   -0.16176,    0.00000,   -0.24925,    0.01937,    0.00000,    0.00000,    0.01253,    0.16127,    0.19061,    0.00000,    0.23933,    1.29131,   -0.01349,    1.00000), 
     (   0.00259,    0.01653,    0.24944,    0.00000,   -0.24698,    0.03875,    0.00000,    0.00000,   -0.03867,   -0.24642,    0.01674,    0.00000,    0.13196,    1.27964,    0.01416,    1.00000), 
     (   0.00000,   -0.25000,    0.00000,    0.00000,   -0.25000,    0.00000,    0.00000,    0.00000,    0.00000,    0.00000,   -0.25000,    0.00000,    0.00234,    1.28078,   -0.01505,    1.00000)]

I didn't *verify* this in all depth, but considering that the `bindShapeMatrix` is

    0.25  0       0     0
    0     0.25    0     0
    0     0       0.25  0
    0     1.29397 0     1

it **strongly** looks like at some point, the `bindShapeMatrix` was not multipled with the `POSITIONS`, but with the `inverseBindMatrices`  (at least, there's some factor 0.25 appearing in many places between the 1.0 and the 2.0 version)

----

Again, all to be taken with a grain of salt, but maybe it helps to find the actual reason of the bug.

**Comment by javagl** [comment]
As discussed in https://github.com/KhronosGroup/COLLADA2GLTF/issues/95#issuecomment-329558429 , it indeed seems like the `bindShapeMatrix` was multiplied with the `inverseBindMatrices` instead of the `POSITIONS`. As soon as this is fixed, regenerating the model should resolve this issue as well. (I'm not sure whether there is a "simple" fix for this, though...)

**Comment by emadurandal** [comment]
I succeeded in salvaging the Maya scene file, which is the base of the WalkingLady model.
（I have not confirmed whether this Maya file completely matches the version of Collada or glTF data adopted as example data on github, but it should be almost the same. In any case, people will also be able to ascertain that by comparing the converted data of this file with the example data adopted.）
 
As shown in the screenshot, this model is characterized by a slight translation and a scale of 0.25 on the mesh.
(I did not make it so intentionally, just because I was not used to making such a skeletal model, I just bound the skeleton to an inadvertently scaled mesh. So it caused confusion this time.)
 
![walkingladymayafile](https://user-images.githubusercontent.com/3821214/31175785-5e92707a-a94b-11e7-9dc9-f0ff906a26a0.jpg)

In my understanding, the role of bindShapeMatrix in glTF 1.0 is to properly map the mesh (raw vertex position data) to the skeleton's coordinate space.
 
Therefore, in the case of glTF 1.0, if WebGL libraries calculates the bone matrices without considering this bindShapeMatrix, a mistake will occur. Skinning will be done on a mesh that is four times as large as the general size (and slightly translated). This factor led to the series of Nightmare phenomena.
 
And in glTF 2.0, I understand that bindShapeMatrix has been removed from the specification and its role has been integrated into inverseBindMatrices. (right?)
Perhaps, in the current glTF 2.0 branch of the COLLADA2GLTF converter, this correspondence to the specification change may be incomplete.
 
Unfortunately I do not have enough skills to find out the potential problems of the COLLADA2GLTF converter. However, With this WalkingLady Maya file, people will be able to create new WalkingLady's glTF 2.0 file in a way other than the COLLADA2GLTF converter.
This will help to check the quality of the current glTF 2.0 branch of the COLLADA2GLTF converter, and (ofcourse) to check wheather the WalkingLady data itself contains bugs or not.
 
Perhaps this "WalkingLady problem" is not dependent on the Maya software because scaling (and etc) are applied at the time of the original modeling data. And with reference to the characteristics of this modeling data, Blender can create a similar model, and similar verification based on Blender may be possible.
[WalkingLady.mb.zip](https://github.com/KhronosGroup/glTF-Sample-Models/files/1355748/WalkingLady.mb.zip)

**Comment by emackey** [comment]
This sample no longer appears in the glTF 2.0 collection.
